test/integration/vhds_integration_test.cc

```diff
@@ -0,0 +1,298 @@
+                                                                   vhds_stream_);
+
+    // Wait for our statically specified listener to become ready, and register its port in the
+    // test framework's downstream listener port map.
+    test_server_->waitUntilListenersReady();
+    registerTestServerPorts({"http"});
+  }
+
+  void useRdsWithVhosts() { use_rds_with_vhosts = true; }
+  const envoy::api::v2::RouteConfiguration rdsConfig() const {
+    return TestUtility::parseYaml<envoy::api::v2::RouteConfiguration>(
+        use_rds_with_vhosts ? RdsConfigWithVhosts : RdsConfig);
+  }
+
+  FakeStreamPtr vhds_stream_;
+  bool use_rds_with_vhosts{false};
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, VhdsIntegrationTest, GRPC_CLIENT_INTEGRATION_PARAMS);
+
+TEST_P(VhdsIntegrationTest, VhdsVirtualHostAddUpdateRemove) {
+  // Calls our initialize(), which includes establishing a listener, route, and cluster.
+  testRouterHeaderOnlyRequestAndResponse(nullptr, 1);
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+
+  sendDeltaDiscoveryResponse<envoy::api::v2::route::VirtualHost>(buildVirtualHost1(), {}, "2",
+                                                                 vhds_stream_);
+  EXPECT_TRUE(
+      compareDeltaDiscoveryRequest(Config::TypeUrl::get().VirtualHost, {}, {}, vhds_stream_));
+  EXPECT_TRUE(
+      compareDeltaDiscoveryRequest(Config::TypeUrl::get().VirtualHost, {}, {}, vhds_stream_));
+
+  testRouterHeaderOnlyRequestAndResponse(nullptr, 1, "/one", "vhost.first");
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+  testRouterHeaderOnlyRequestAndResponse(nullptr, 1, "/two", "vhost.second");
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+
+  sendDeltaDiscoveryResponse<envoy::api::v2::route::VirtualHost>({}, {"vhost_1", "vhost_2"}, "3",
+                                                                 vhds_stream_);
+  EXPECT_TRUE(
+      compareDeltaDiscoveryRequest(Config::TypeUrl::get().VirtualHost, {}, {}, vhds_stream_));
+  EXPECT_TRUE(
+      compareDeltaDiscoveryRequest(Config::TypeUrl::get().VirtualHost, {}, {}, vhds_stream_));
+
+  codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+  Http::TestHeaderMapImpl request_headers{{":method", "GET"},
+                                          {":path", "/one"},
+                                          {":scheme", "http"},
+                                          {":authority", "vhost.first"},
+                                          {"x-lyft-user-id", "123"}};
+  IntegrationStreamDecoderPtr response = codec_client_->makeHeaderOnlyRequest(request_headers);
+  response->waitForHeaders();
+  EXPECT_STREQ("404", response->headers().Status()->value().c_str());
+
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(VhdsIntegrationTest, RdsWithVirtualHostsVhdsVirtualHostAddUpdateRemove) {
```

Comment:
Can you add single line `//` human readable description above each `TEST_P`?

Reply:
Added